src/urg_sim/urg_sim.cpp

```diff
@@ -0,0 +1,761 @@
+{
+  if (!validateExtraString(cmd, 2))
+  {
+    return;
+  }
+  // Two RB commands within one second triggers sensor reboot
+  const auto now = boost::posix_time::microsec_clock::universal_time();
+  if (last_rb_ == boost::posix_time::not_a_date_time ||
+      now - last_rb_ > boost::posix_time::seconds(1))
+  {
+    response(cmd, "01");
+    last_rb_ = now;
+    return;
+  }
+  response(cmd, status_ok);
+
+  boot_timer_.expires_from_now(boost::posix_time::seconds(1));
+  boot_timer_.async_wait(
+      boost::bind(
+          &URGSimulator::reboot,
+          this));
+}
+
+void URGSimulator::handleMX(const std::string cmd)
+{
+  if (!validateExtraString(cmd, 15))
+  {
+    return;
+  }
+
+  std::lock_guard<std::mutex> lock(mu_);
+
+  try
+  {
+    measurement_start_step_ = std::stoi(cmd.substr(2, 4));
+    measurement_end_step_ = std::stoi(cmd.substr(6, 4));
+    measurement_grouping_step_ = std::stoi(cmd.substr(10, 2));
+    measurement_skips_ = std::stoi(cmd.substr(12, 1));
+    measurement_scans_ = std::stoi(cmd.substr(13, 2));
+  }
+  catch (const std::invalid_argument& e)
+  {
+    response(cmd, "02");
+    return;
+  }
+  if (measurement_start_step_ > measurement_end_step_)
+  {
+    response(cmd, "05");
+    return;
+  }
+  if (measurement_start_step_ < params_.angle_min ||
+      measurement_end_step_ > params_.angle_max)
+  {
+    response(cmd, "04");
+    return;
+  }
+  if (measurement_grouping_step_ == 0)
+  {
+    measurement_grouping_step_ = 1;
+  }
+  measurement_cnt_ = 0;
+  measurement_sent_ = 0;
+  measurement_cmd_ = cmd.substr(0, 13);
+  measurement_extra_string_ = cmd.substr(15);
+
+  sensor_state_ = SensorState::MULTI_SCAN;
+  switch (cmd[1])
+  {
+    case 'D':
+      measurement_mode_ = MeasurementMode::RANGE;
+      break;
+    case 'E':
+      measurement_mode_ = MeasurementMode::RANGE_INTENSITY;
+      break;
+  }
+  response(cmd, status_ok);
+}
+
+void URGSimulator::handleUnknown(const std::string cmd)
+{
+  if (cmd == "")
+  {
+    return;
+  }
+  response(cmd, status_error_command_not_defined);
+}
+
+void URGSimulator::handleDisconnect()
+{
+  std::lock_guard<std::mutex> lock(mu_);
+
+  if (sensor_state_ != SensorState::BOOTING &&
+      sensor_state_ != SensorState::ERROR_DETECTED)
+  {
+    sensor_state_ = SensorState::IDLE;
+  }
+  socket_.close();
+  accept();
+}
+
+void URGSimulator::reboot()
+{
+  std::cerr << "Booting" << std::endl;
+  {
+    std::lock_guard<std::mutex> lock(mu_);
+    sensor_state_ = SensorState::BOOTING;
+  }
+
+  const auto delay = boost::posix_time::microseconds(
+      static_cast<int64_t>(params_.boot_duration * 1e6));
+  boot_timer_.expires_from_now(delay);
+  boot_timer_.async_wait(
+      boost::bind(
+          &URGSimulator::booted,
+          this));
+  timestamp_epoch_ = boost::posix_time::microsec_clock::universal_time();
+
+  if (socket_.is_open())
+  {
+    std::cerr << "Closing" << std::endl;
+    socket_.close();
+    accept();
+  }
+  accept();
```

Comment:
Is this intended to call `accept()` twice if `socket_.is_open() == true`?

Reply:
Removed duplicated `accept`